cmd/go-infrabin/main.go

```diff
@@ -1,176 +1,31 @@
-		}
-		time.Sleep(time.Duration(helpers.Min(seconds, maxDelay)) * time.Second)
-
-		resp.Delay = strconv.Itoa(seconds)
-		data := helpers.MarshalResponseToString(resp)
-
-		w.WriteHeader(http.StatusOK)
-		_, err = io.WriteString(w, data)
-		if err != nil {
-			log.Fatal("error writing to ResponseWriter", err)
-		}
-	}
-}
-
-// HeadersHandler handles the headers endpoint
-func HeadersHandler(w http.ResponseWriter, r *http.Request) {
-	var resp helpers.Response
-	w.Header().Set("Content-Type", "application/json")
-	resp.Headers = &r.Header
-
-	data := helpers.MarshalResponseToString(resp)
-	_, err := io.WriteString(w, data)
-	if err != nil {
-		log.Fatal("error writing to ResponseWriter: ", err)
-	}
-}
-
-// EnvHandler handles the env endpoint
-func EnvHandler(w http.ResponseWriter, r *http.Request) {
-	var resp helpers.Response
-	vars := mux.Vars(r)
-	w.Header().Set("Content-Type", "application/json")
-	value := helpers.GetEnv(vars["env_var"], "")
-
-	if value == "" {
-		data := helpers.MarshalResponseToString(resp)
-		w.WriteHeader(http.StatusNotFound)
-		_, err := io.WriteString(w, data)
-		if err != nil {
-			log.Fatal("error writing to ResponseWriter", err)
-		}
-	} else {
-
-		resp.Env = map[string]string{
-			vars["env_var"]: value,
-		}
-		data := helpers.MarshalResponseToString(resp)
-
-		w.WriteHeader(http.StatusOK)
-		_, err := io.WriteString(w, data)
-		if err != nil {
-			log.Fatal("error writing to ResponseWriter", err)
-		}
-
-	}
-}
 
 func main() {
-	r := mux.NewRouter()
-	a := mux.NewRouter()
-	finish := make(chan bool)
-
-	r.HandleFunc("/", RootHandler)
-	r.HandleFunc("/delay/{seconds}", DelayHandler)
-	r.HandleFunc("/headers", HeadersHandler)
-	r.HandleFunc("/env/{env_var}", EnvHandler)
-
-	a.HandleFunc("/liveness", LivenessHandler)
-
-	serviceSrv := &http.Server{
-		Handler: r,
-		Addr:    "0.0.0.0:8888",
-		// Good practice: enforce timeouts
-		WriteTimeout: 15 * time.Second,
-		ReadTimeout:  15 * time.Second,
-	}
-	adminSrv := &http.Server{
-		Handler: a,
-		Addr:    "0.0.0.0:8899",
-		// Good practice: enforce timeouts
-		WriteTimeout: 15 * time.Second,
-		ReadTimeout:  15 * time.Second,
-	}
-
-	log.Print("starting go-infrabin")
-
-	go func() {
-		log.Print("Listening on service port 8888...")
-		log.Fatal(serviceSrv.ListenAndServe())
-	}()
-
-	go func() {
-		log.Print("Listening on admin port 8899...")
-		log.Fatal(adminSrv.ListenAndServe())
-	}()
+	// Create a channel to catch signals
+	finish := make(chan os.Signal, 1)
+	// We'll accept graceful shutdowns when quit via SIGINT (Ctrl+C)
+    // SIGKILL, SIGQUIT or SIGTERM (Ctrl+/) will not be caught.
+	signal.Notify(finish, os.Interrupt)
+	
+	// run service server in background
+	server := infrabin.NewHTTPServer()
+	go server.ListenAndServe()
+
+	// run admin server in background
+	admin := infrabin.NewAdminServer()
+	go admin.ListenAndServe()
+
+	// wait for SIGINT
 	<-finish
+
+	admin.Shutdown()
+	server.Shutdown()
```

Comment:
This does mean that it shuts down one then the other.
Need better `context` foo.

Reply:
Not sure what you mean here. What we should do?

Reply:
I would say it is ok for now.